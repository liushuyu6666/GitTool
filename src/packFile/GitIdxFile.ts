import { readFileSync } from 'fs';
import path from 'path';
import { GitFanout } from './GitFanout';

export interface GitIdxFileInterface {
  byteSize: number;

  header: number[];

  versionNumber: number;

  fanout: GitFanout;

  parseHeader(content: Buffer): [number, number, number, number];

  parseVersionNumber(content: Buffer): number

}

export class GitIdxFile implements GitIdxFileInterface {
  // size by bytes
  byteSize: number;

  header: number[];

  versionNumber: number;

  fanout: GitFanout;

  constructor(filePath: string, outDir: string) {
    const fileName = filePath.substring(filePath.lastIndexOf(path.sep));
    const outFile = path.join(outDir, fileName);
    const content = readFileSync(filePath);
    this.byteSize = content.length;
    this.header = this.parseHeader(content);
    this.versionNumber = this.parseVersionNumber(content);
    this.fanout = new GitFanout(content, outFile);
  }

  parseHeader(content: Buffer): [number, number, number, number] {
    // [0, 3] bytes
    return [
      content.readUInt8(0),
      content.readUInt8(1),
      content.readUInt8(2),
      content.readUInt8(3),
    ];
  }

  parseVersionNumber(content: Buffer): number {
    return content.readUInt32BE(4);
  }
}
