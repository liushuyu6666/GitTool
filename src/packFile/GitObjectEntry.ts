import { ManipulateBuffer } from "../manipulateBuffer/ManipulateBuffer";
import getGitObjectType, { GitObjectType } from "../utils/getGitObjectType";

export interface GitObjectEntryInterface {
  type: GitObjectType;

  size: number;

  bodyStartIndex: number;

  bodyEndIndex: number;

  getSize(chunk: Buffer): [number, number];
}

export class GitObjectEntry implements GitObjectEntryInterface {
  type: GitObjectType;

  size: number;

  bodyStartIndex: number;

  bodyEndIndex: number;

  constructor(content: Buffer, startIndex: number, endIndex: number) {
    const chunk = content.subarray(startIndex, endIndex);

    const typeNumber = (chunk[0] & 0b01110000) >> 4 & 0b00000111;

    this.type = getGitObjectType(typeNumber);

    [this.size, this.bodyStartIndex] = this.getSize(chunk);

    this.bodyEndIndex = endIndex;
  }

  // TODO: ManipulateBuffer should be recycled.
  getSize(chunk: Buffer): [number, number] {
    const manipulateBuffer = new ManipulateBuffer(false);
    let msb: boolean = chunk[0] > 0b10000000;
    let index = 1;
    manipulateBuffer.fill(chunk[0], 4, 8);
    while (msb) {
      manipulateBuffer.fill(chunk[index], 1, 8);
      msb = chunk[index] > 0b10000000
      index++;
    }
    manipulateBuffer.finish();

    return [manipulateBuffer.readIntBE(), index];
  }
}