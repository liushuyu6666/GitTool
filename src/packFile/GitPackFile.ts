import { readFileSync } from 'fs';
import { Offset } from './GitFanout';
import { GitObjectEntry } from './GitObjectEntry';

// export interface PackInfo {
//   type: number;
//   info: string;
// }

export interface GitPackInterfaceFile {
  size: number;

  // 'PACK'
  layer1: string;

  // version
  layer2: number;

  // object hex list
  layer3: number;

  // data chunk
  layer4: Record<string, GitObjectEntry>;

  // checksum
  layer5: Buffer;

  // layer 1: [0, 4)
  parseLayer1(content: Buffer): string;
  // layer 2: [4, 8)
  parseLayer2(content: Buffer): number;

  // layer 3: [8, 12)
  parseLayer3(content: Buffer): number;

  parseLayer4(offsets: Record<string, number>, content: Buffer): Record<string, GitObjectEntry>;

  parseLayer5(content: Buffer): Buffer;
}

// TODO: need setter and getter
export class GitPackFile implements GitPackInterfaceFile {
  size: number;

  // 'PACK'
  layer1: string;

  // version
  layer2: number;

  // object hex list
  layer3: number;

  // data chunk
  layer4: Record<string, GitObjectEntry>;

  // checksum
  layer5: Buffer;

  constructor(filePath: string, offsets: Record<string, number>) {
    const content = readFileSync(filePath);
    this.size = content.length;
    this.layer1 = this.parseLayer1(content);
    this.layer2 = this.parseLayer2(content);
    this.layer3 = this.parseLayer3(content);
    this.layer4 = this.parseLayer4(offsets, content);
    this.layer5 = this.parseLayer5(content);
  }

  // layer 1: [0, 4)
  parseLayer1(content: Buffer): string {
    return content.subarray(0, 4).toString('utf-8');
  }

  // layer 2: [4, 8)
  parseLayer2(content: Buffer): number {
    return content.subarray(4, 8).readUInt32BE();
  }

  // layer 3: [8, 12)
  parseLayer3(content: Buffer): number {
    return content.subarray(8, 12).readUInt32BE();
  }

  parseLayer4(offsets: Record<string, number>, content: Buffer): Record<string, GitObjectEntry> {
    // create offsetArray
    const offsetArray: Offset[] = [];
    Object.entries(offsets).forEach(([hex, offset]) => {
      offsetArray.push({
        hex,
        offset
      })
    })

    // sort offsetArray ascending
    // offsetArray.length = offsets.length + 1
    offsetArray.sort((prev, next) => prev.offset - next.offset);
    offsetArray.push({
      hex: '',
      offset: content.length - 20 // this is the endIndex of the last object entry
    })

    const gitPackObjectEntry: Record<string, GitObjectEntry> = {};
    for (let i = 0; i < Object.keys(offsets).length; i++) {
      const startIndex = offsetArray[i].offset;
      const endIndex = offsetArray[i + 1].offset;
      const hex = offsetArray[i].hex;
      const entry = new GitObjectEntry(content, startIndex, endIndex);
      gitPackObjectEntry[hex] = entry;
    }
    return gitPackObjectEntry;
  }

  parseLayer5(content: Buffer): Buffer {
    return content.subarray(content.length - 20);
  }
}
