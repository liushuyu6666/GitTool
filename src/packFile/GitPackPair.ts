import { GitObject } from '../gitObject/GitObject';
import { isPackDeltaObject, isPackOriginalObject } from '../utils/getGitObjectType';
import { GitIdxFile } from './GitIdxFile';
import { GitPackFile } from './GitPackFile';

export interface GitPackPairInterface {
  pack: GitPackFile;

  idx: GitIdxFile;
}

export interface GitPackObjects {
  packOriginalObjects: GitObject[];
  packDeltaObjects: GitObject[];
}

export class GitPackPair implements GitPackPairInterface {
  pack: GitPackFile;

  idx: GitIdxFile;

  filePathIdx: string;

  filePathPack: string;

  constructor(filePathIdx: string, filePathPack: string, outPackPath: string) {
    this.filePathIdx = filePathIdx;
    this.filePathPack = filePathPack;
    this.idx = new GitIdxFile(filePathIdx, outPackPath);
    this.pack = new GitPackFile(filePathPack, this.idx.fanout.offsets);
  }

  generateGitPackObjects(): GitPackObjects {
    const packOriginalObjects: GitObject[] = [];
    const packDeltaObjects: GitObject[] = [];

    for (const [hash, gitObjectEntry] of Object.entries(this.pack.layer4)) {
      const type = gitObjectEntry.type;

      if (isPackOriginalObject(type)) {
        packOriginalObjects.push(
          new GitObject({
            hash,
            type,
            size: gitObjectEntry.size,
            filePath: this.filePathPack,
            bodyOffsetStartIndex: gitObjectEntry.bodyStartIndex,
            bodyOffsetEndIndex: gitObjectEntry.bodyEndIndex,
          }),
        );
      } else if (isPackDeltaObject(type)) {
        packDeltaObjects.push(
          new GitObject({
            hash,
            type,
            size: gitObjectEntry.size,
            filePath: this.filePathPack,
            bodyOffsetStartIndex: gitObjectEntry.bodyStartIndex,
            bodyOffsetEndIndex: gitObjectEntry.bodyEndIndex,
          }),
        );
      }
    }

    return {
      packOriginalObjects,
      packDeltaObjects,
    };
  }

  // generateGitOriginalObject(): GitObject[] {
  //   const gitObjects: GitObject[] = [];

  //   for (const [hash, gitObjectEntry] of Object.entries(this.pack.layer4)) {
  //     const type = gitObjectEntry.type;

  //     if (isOriginalDeltaObject(type)) {
  //       gitObjects.push(
  //         new GitObject({
  //           hash,
  //           type,
  //           size: gitObjectEntry.size,
  //           filePath: this.filePathPack,
  //           bodyOffsetStartIndex: gitObjectEntry.bodyStartIndex,
  //           bodyOffsetEndIndex: gitObjectEntry.bodyEndIndex,
  //         }),
  //       );
  //     }
  //   }

  //   return gitObjects;
  // }

  // TODO: maybe we can combine generateGitDeltaObject with generateGitOriginalObject
  // generateGitDeltaObject(): GitObject[] {
  //   const gitObjects: GitObject[] = [];

  //   for (const [hash, gitObjectEntry] of Object.entries(this.pack.layer4)) {
  //     const type = gitObjectEntry.type;

  //     if (isDeltaObject(type)) {
  //       gitObjects.push(
  //         new GitObject({
  //           hash,
  //           type,
  //           size: gitObjectEntry.size,
  //           filePath: this.filePathPack,
  //           bodyOffsetStartIndex: gitObjectEntry.bodyStartIndex,
  //           bodyOffsetEndIndex: gitObjectEntry.bodyEndIndex,
  //         }),
  //       );
  //     }
  //   }

  //   return gitObjects;
  // }
}
