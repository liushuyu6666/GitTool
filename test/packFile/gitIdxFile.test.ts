import { GitIdxFile } from '../../src/packFile/GitIdxFile';

describe('Test GitIdx Class', () => {
  const filePath =
    'gitTest/objects/pack/pack-5eb139577548314dffcb8e6410b30413a81ca3fb.idx';
  const gitIdx = new GitIdxFile(filePath);

  test('test header, should be [255, 116, 79, 99]', () => {
    expect(gitIdx.header).toEqual([255, 116, 79, 99]);
  });

  test('test version number, should be 2', () => {
    expect(gitIdx.versionNumber).toBe(2);
  });

  test('test entry size of fanout table', () => {
    const entrySize = gitIdx.fanout.entrySize;

    expect(entrySize).toBe(167);
  });

  test('test layer1 of fanout table', () => {
    const layer1 = gitIdx.fanout.layer1;
    const totalEntries = Object.values(layer1).reduce(
      (prev, curr) => prev + curr,
      0,
    );
    expect(layer1).toMatchSnapshot('layer1');
    expect(totalEntries).toBe(167);
  });

  test('test offsets of fanout table', () => {
    const offsets = gitIdx.fanout.offsets;

    expect(offsets).toMatchSnapshot('offsets');
  });
});
