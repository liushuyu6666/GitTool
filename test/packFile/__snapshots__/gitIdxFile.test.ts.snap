// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GitIdx Class test layer1 of fanout table: layer1 1`] = `
{
  "0": 0,
  "1": 0,
  "10": 0,
  "100": 0,
  "101": 1,
  "102": 4,
  "103": 0,
  "104": 0,
  "105": 0,
  "106": 1,
  "107": 1,
  "108": 0,
  "109": 0,
  "11": 0,
  "110": 0,
  "111": 1,
  "112": 0,
  "113": 1,
  "114": 2,
  "115": 0,
  "116": 0,
  "117": 0,
  "118": 1,
  "119": 1,
  "12": 0,
  "120": 1,
  "121": 0,
  "122": 0,
  "123": 0,
  "124": 0,
  "125": 1,
  "126": 2,
  "127": 0,
  "128": 2,
  "129": 0,
  "13": 2,
  "130": 1,
  "131": 2,
  "132": 1,
  "133": 2,
  "134": 0,
  "135": 1,
  "136": 0,
  "137": 0,
  "138": 0,
  "139": 0,
  "14": 0,
  "140": 0,
  "141": 0,
  "142": 0,
  "143": 1,
  "144": 0,
  "145": 1,
  "146": 3,
  "147": 0,
  "148": 0,
  "149": 2,
  "15": 1,
  "150": 0,
  "151": 1,
  "152": 0,
  "153": 0,
  "154": 1,
  "155": 0,
  "156": 2,
  "157": 0,
  "158": 0,
  "159": 2,
  "16": 1,
  "160": 1,
  "161": 0,
  "162": 0,
  "163": 0,
  "164": 1,
  "165": 1,
  "166": 2,
  "167": 1,
  "168": 0,
  "169": 1,
  "17": 1,
  "170": 0,
  "171": 1,
  "172": 3,
  "173": 0,
  "174": 1,
  "175": 0,
  "176": 1,
  "177": 1,
  "178": 1,
  "179": 0,
  "18": 0,
  "180": 0,
  "181": 0,
  "182": 1,
  "183": 2,
  "184": 0,
  "185": 0,
  "186": 0,
  "187": 1,
  "188": 1,
  "189": 2,
  "19": 0,
  "190": 2,
  "191": 1,
  "192": 0,
  "193": 1,
  "194": 1,
  "195": 0,
  "196": 0,
  "197": 0,
  "198": 0,
  "199": 0,
  "2": 1,
  "20": 1,
  "200": 0,
  "201": 1,
  "202": 1,
  "203": 0,
  "204": 0,
  "205": 0,
  "206": 1,
  "207": 1,
  "208": 0,
  "209": 2,
  "21": 0,
  "210": 1,
  "211": 2,
  "212": 1,
  "213": 0,
  "214": 0,
  "215": 1,
  "216": 1,
  "217": 0,
  "218": 0,
  "219": 1,
  "22": 0,
  "220": 4,
  "221": 0,
  "222": 0,
  "223": 1,
  "224": 1,
  "225": 1,
  "226": 2,
  "227": 0,
  "228": 1,
  "229": 1,
  "23": 1,
  "230": 0,
  "231": 1,
  "232": 0,
  "233": 4,
  "234": 0,
  "235": 2,
  "236": 0,
  "237": 1,
  "238": 1,
  "239": 0,
  "24": 3,
  "240": 2,
  "241": 1,
  "242": 1,
  "243": 0,
  "244": 1,
  "245": 0,
  "246": 0,
  "247": 0,
  "248": 0,
  "249": 0,
  "25": 0,
  "250": 0,
  "251": 1,
  "252": 3,
  "253": 2,
  "254": 0,
  "255": 0,
  "26": 0,
  "27": 0,
  "28": 1,
  "29": 0,
  "3": 0,
  "30": 1,
  "31": 1,
  "32": 0,
  "33": 0,
  "34": 0,
  "35": 0,
  "36": 0,
  "37": 0,
  "38": 1,
  "39": 1,
  "4": 2,
  "40": 0,
  "41": 0,
  "42": 0,
  "43": 0,
  "44": 0,
  "45": 1,
  "46": 0,
  "47": 2,
  "48": 0,
  "49": 1,
  "5": 1,
  "50": 1,
  "51": 0,
  "52": 0,
  "53": 2,
  "54": 1,
  "55": 0,
  "56": 1,
  "57": 1,
  "58": 0,
  "59": 0,
  "6": 0,
  "60": 1,
  "61": 1,
  "62": 1,
  "63": 0,
  "64": 1,
  "65": 2,
  "66": 2,
  "67": 0,
  "68": 0,
  "69": 2,
  "7": 0,
  "70": 1,
  "71": 0,
  "72": 0,
  "73": 0,
  "74": 1,
  "75": 1,
  "76": 0,
  "77": 0,
  "78": 1,
  "79": 1,
  "8": 0,
  "80": 1,
  "81": 1,
  "82": 1,
  "83": 2,
  "84": 0,
  "85": 0,
  "86": 1,
  "87": 0,
  "88": 0,
  "89": 0,
  "9": 0,
  "90": 1,
  "91": 2,
  "92": 0,
  "93": 0,
  "94": 0,
  "95": 1,
  "96": 1,
  "97": 0,
  "98": 0,
  "99": 2,
}
`;

exports[`Test GitIdx Class test offsets of fanout table: offsets 1`] = `
{
  "02a50eacdaeb0b86aaa442edfc9ee3835b5f0e7b": 34527,
  "043d1a7a5ba42f3cd0c8d6a7009c02796d281316": 35910,
  "04b56df968e4208fd518a2f1bb7b923813d33da9": 86081,
  "0571fcd97f01980324e3f9136d321e9d494d94e6": 16767,
  "0d7c8aa9055c1d06d2eafa3fa5abb2bf75be3d04": 24574,
  "0d9024bc7f7270aee8aca4a1c596a325f93860a3": 24859,
  "0f3b5bcad451971ffb0b21c69f56f883b008398d": 24801,
  "10cbf509fa8a0c375b9f6b3777cab0226c670819": 94237,
  "119f2df64472dc761814a133e2695057970c0c1f": 9288,
  "14fede6a43a129a6832034c54cf400659c25062d": 86558,
  "1795d9a1428b09f9feef592c5395e17f01ac9bc7": 72601,
  "181644dde8d586a91d6b881ccb49347df337480b": 9239,
  "183b2873afeca2e1f1ec7fe911a7696a2e5d5cf6": 3563,
  "1862fe809350eb4d7a9fdf38bc0f8a65172dfbd4": 88199,
  "1cd88f3f9596100439f46e456b45db5f31062d66": 20714,
  "1ee56c1d6749dcd66cc80c2d66238def3279579d": 39543,
  "1f31f54b0eec6c91eff0ce5cbcc8761fa92b2bf3": 28553,
  "265b7b474916a7099bb36fdb3f195d576b213807": 28015,
  "2705b5bc2fccea0856b621aa4f7cee4b0493f608": 36836,
  "2d4d46f84bd32e00d44d8a86d1c03d13038fd118": 24230,
  "2faae8f2b8f9ee9b81469a20716d021d1fb3b0e8": 12806,
  "2fdb4e1407f117455e6225d647aa2c41a31fcbf2": 27947,
  "317d38c07ae453ada3cb8eeca73de5b1ce1fb681": 40566,
  "32161f4e1d6dfd0476fb477b7682821adb415490": 21243,
  "3500099b89a097a956e0f06d0695d5cd626ff058": 25643,
  "359250edf9ec653b985a4635b5578f92d4b597a6": 66951,
  "36a3a0926b7292d79de9baa94fd645235b8ee37b": 12412,
  "384ff77394e2c73b9af23d0da581e55f7df64186": 37888,
  "39ab1ddec31bd5e8b23febccafe4806e336c4f4a": 27367,
  "3c63802ac527ab92f0fc87848c6e934a13cd2d86": 42759,
  "3db9f984dfee2e6cb4b8ddf863d54db7aed98582": 15166,
  "3e7afed771eb0eeff37bdbde1ae476ec316ff1cf": 86968,
  "409080c241792233cea77b49c269947b851c0331": 83071,
  "411113ee43c7ff668ee73b1056c2b8b29d0125bf": 41351,
  "41c9381dab68772d7136e6749aa1c51f5a398275": 18207,
  "42305fdaf48e40da8b34cb12d037cbf4870063b9": 24917,
  "4275228b4d7a95b2a3615b376810fa40cd8bbc37": 16866,
  "450c0a46594a63646cc4ee33706053db6e8d1a41": 4227,
  "45254c8eab38e9e147b9babaf7c201cc01a9538d": 16231,
  "46af2631fd1bc353e90a5ab5533e7b260617ec51": 13015,
  "4a6092c249bfbca806daa86d8ce283cb306f735e": 34409,
  "4b8f61eb2ddc2a8fbe884253982c1b640ff35e28": 19989,
  "4e5b18bc1d4421a5cace974b786abe318b7d106d": 68029,
  "4feeaebd32b33ca57f5e268f793dea81e0a60324": 1177,
  "506c71d6fa1a8d19d8ad22bad9d8cf336bb894c0": 28696,
  "51cac3c29f4d484fb24ea4f8870ee31ddb95e6f0": 47219,
  "5225c320ff02f00c35944ce359af29916864a21c": 25416,
  "53c4fa62dcac6e2a7a118eaec119d2e1cd5e0f47": 17000,
  "53f414b29e97c7ba99a7099d125c8777f2b8a11d": 33326,
  "56a65082c2c3652ba47dec00c749f3594280afb4": 188,
  "5a7064bedf1c92de495375782ebd6d084de0a45a": 25731,
  "5b99f0cbb0276e5134818443f30c4c0b1d13d797": 34588,
  "5bd610490c625157912c8ea54845bb8a08f24798": 37972,
  "5f1562f52e1dfc8c0851d0d5502370e68dae2134": 25324,
  "60d8551e56d5610e5b0632ca6725ba7c53493583": 16436,
  "635a0dc14ec065b5b6e8f9fbc534f64a3386ed4b": 9189,
  "639d2a62d780fa8a84f4b9171611a02869f99301": 74937,
  "65cd2ff82bc4d417a554c4b6133007d7ef7f2973": 1608,
  "660afd62f3722d838990ad92cc54e402d52254d8": 21068,
  "667acc652596a6e4496ae60b84634fcf39482943": 2310,
  "66ba6f0fdc5e93073b1e46206d5ca89bc6e8e766": 33372,
  "66e6611a62eeeeae558510f45982631a601308af": 15304,
  "6a4b909622b516eccb734eb4b889729d3068428d": 38019,
  "6b289e7f4c7d23998c1fa40f48c4d788d30df007": 27557,
  "6f39c6739f1f1d1e4e3ac2870913321be1b7d420": 6535,
  "719983047892ae8107c71545fb52e2a13b040726": 27320,
  "7211c8e0559195248b3d8379a8e0dd83546d0185": 12536,
  "72a783f81641e31bd057a66d542d22983e513f45": 15518,
  "76c301d8977ba7f0cef1a39748c260375306cd8b": 7507,
  "77f4296d42747076ba732726494077aa4374259e": 3768,
  "78fe90efda83c115291608dfab8b46c7f89e1b7d": 55704,
  "7d00cdc98447b40d150030f59b788d1b127c7c95": 15235,
  "7e6caf1e107ca80dae2ad20b1a003c0c46ad7ac7": 19413,
  "7ed40f78115acfefe2a4d22313ad0a4b62a7c658": 15824,
  "800139a25d8391d47bbe2ca434391260e802f43e": 12,
  "806525178154bec6cfd7571c5307bdc699b24ed5": 16697,
  "82ed7fae4a75dec313af3a5ee16958faac5084ea": 6977,
  "83339e531cc60d6c61aeb3600e0b7eedf9bc194d": 5194,
  "83b8e96785ee8ed39eeaec1b540e0263ca6b3cab": 33279,
  "849cd14ee2a72e5319f86e64c80e60b06c0467e2": 36184,
  "8557fcab609cf165a7fb8748759f276fb28b842d": 28939,
  "85df988beffc4f85c30325c81f79b883e3504c3d": 27902,
  "874e8569d84776aeffbee979e15d1e94b6b0fe74": 20166,
  "8f884ab80f50a44889281a0f207abc70ef176775": 22721,
  "915ada98a651c38687d4bc04b0d4eb032feacdc6": 13285,
  "9234881e4c1668cfc20c47beca73524a4eba0004": 63832,
  "925707b48b204d18f60f92fb2825a42264a1df9d": 2724,
  "92c2d53330515b38b6326c9a0260f101b3485732": 25461,
  "951ea58ade28f29e61ba04cfd3c1018dcdb3ce5b": 28088,
  "958a8681324ca8c7daba53f11727e5ad2c7c2c2c": 25168,
  "97ccda240c65f8e61c8b4dc70dbcfa1d977e633a": 35880,
  "9a7616aacbd66148182b0175d54369f2f9f74758": 10474,
  "9c2957bf22418f2a0c115a87808a6e158e193813": 25807,
  "9c834d735c3670b24b9d1cdef61aebc16ff95d68": 1000,
  "9f2c649b9e2010c64b2b2df4b0c5bf56e6109e82": 33415,
  "9f6435680ab54a7cbfd9753be3784dac4ce49188": 93743,
  "a0766d8c88b8045f7c8df9cafe99f534f12cdfcd": 1318,
  "a43eb2703663692fd971f5ad8d1646515ed2a801": 33064,
  "a5b1a7b677778f38dceecfdf9c1a57e6ba8fe5d9": 37859,
  "a6022fdf0ba005a13e278feee61ba725325e0671": 35667,
  "a6a739b453059a782297796c865885f64aff7486": 28391,
  "a73f6bbd7450e49a165fa5a484b4418eeebec016": 17601,
  "a94be8d1ecbf05dfd66466ca2973a8060fb75809": 47583,
  "ab9662b7862222891e80bbd9158f2260ac19f7fe": 12034,
  "ac126b319b2a860c0595c3ff66bc644783326413": 90943,
  "accc1792b3b1b05340df129d88a8473bbdf60ab5": 85557,
  "acd8de2e36e46e2e6b36f9037aa0d632662db61d": 28438,
  "ae7bf1dbcaa9b2701766bac1a9139c27270f9b77": 23791,
  "b0ae0c71bec7feec75113c09261bb467a4f187e9": 37527,
  "b1869d2bf16c75e9ccaf4ba5b70194817f02cfa4": 28853,
  "b2cd371bec6ad64593ad0cf213539653fb7091a3": 46346,
  "b657127060104ad0ce6e41b2c94da1eb39086b12": 13113,
  "b70f858b6d7884cc0078a3a8e90d92009032367d": 7552,
  "b7a2baa9b15efbf7d07a3c019d537e35e8d01212": 5669,
  "bb28ede5636073f8f90ebc83d4f0de4924872d9d": 40378,
  "bcca21c6738332d722bca338da5c95c6072c282a": 33005,
  "bd36efa932eb1830a9b992c2d2d70b6c9581971a": 28811,
  "bd98fab5d804016b72f0ab1a2caf5abc03415838": 25279,
  "be0b8162d306d6991ba4def3167e1ccff94a88f4": 38064,
  "beeff05c3d838c9dfdd31de36694febff5d89e76": 12719,
  "bf4ea68b160706dd05f2870629327fbfacba4669": 9392,
  "c1400b3f221abdfd4df7b8005e2a56243888054b": 4330,
  "c2dcc56ae0f468d7c27b53695d192678650b7e2a": 41127,
  "c9bf495f5bcfa2dc913c7aca856830594b17cd72": 46058,
  "ca13fb1db3b312a5e2ad55d8000d8df7df2c1a80": 12592,
  "ce115aaaca8e743607b9e5bd253331732dd5eb93": 4123,
  "cf12ec5aeec16414726f5df188e1f76c895be657": 4019,
  "d18896298e09d8fe2684a2b25c882992c2b170a5": 20891,
  "d1a1f742c45550d3a5fb3947ec17749b5a8f267e": 15759,
  "d2f05796ccb8c3e732257e026485af73dfc78315": 21384,
  "d3235a6929522e51ad6a35d40a1932ea2937ef10": 13193,
  "d34ce4e127cc191a928baa4205944483c7be7e20": 11954,
  "d442b1fb3d472042916d33ee5f0e82477b5122bf": 8970,
  "d703aed5cd52c9b01fdbfa94d6f7b09651ba053b": 28600,
  "d81403cffb9a2c720d398cd7bccebbc1d38ae3b5": 36247,
  "db34cb0448242e2263a55dbd38072490e62e0f47": 7324,
  "dc3bec3d91f94893a1faedcf638e82eb88c4aec2": 16107,
  "dc5a4cd442368f4d07e07f3ef52aebcfa2e3c3a5": 22247,
  "dc980949ec3f2a3160c6f1542c39b38f4abbbb05": 38114,
  "dce87b5b8556d87b2b8630de9d594d4202249816": 18826,
  "df8f5676be476dbbd0f6637da8d4208b56147c97": 94044,
  "e05edbff809d459dd9f6f2829445cd01d2db72fc": 27635,
  "e18ea8abf0043f31988d8fea36dbd82a05980ddc": 27417,
  "e23b552d15d0ddce8ff68124a0bbc2cc99cf5e11": 43774,
  "e2ac4c4f890df4e15e941b85ed3af4c9c1d2c1ab": 12653,
  "e4280495b8adf9a27d832bb9ff57ce0779b9315b": 28346,
  "e5ee6f8718deef8c88c214a62f11af41253e5d6b": 27228,
  "e7b295253cde4c267097f6d795347b4e2c5bd4ad": 41041,
  "e900ea8510599c25600355bd772d200a8e8b6194": 27489,
  "e91a71684d6d7ad42e0df3dc070b4387a6a5948e": 37929,
  "e9c16fc6e495cc8903488ad2b82fa67030c56309": 12349,
  "e9d3390127d4fba23267a4c73211a09d42d1a19f": 25371,
  "eb383b6d93b4927f248f43d9a4ad2b774c615fa0": 76495,
  "eb9a0464fff2492dafc9f6e10a362de295d69740": 23611,
  "ed128cdc111e492f5614e2d0aa960122d4407c79": 45284,
  "ee4be26311770568c2637fced41e3342e8038cff": 16945,
  "f0577776fe0215d3aa03db121cf5a3cc812b856d": 11996,
  "f0f37938046cac7c5f69f17733be84cf14060ca9": 28645,
  "f1c9ab803d5ca193bc9cc51806c83e646b609353": 12475,
  "f22eaacd4471e36fb6308493ed9e219cd5325beb": 23688,
  "f47d2387bbeeb9682432fd1becc1fed581c1891f": 12898,
  "fb7328fe6c628e7c73a6f5cf734a461e5e5e1a0f": 37482,
  "fc9169de9f9107d1c2e3f44f7449fffd9d3c6acb": 27791,
  "fcac36a911157c21a50c0609fe932b87cc0d1036": 36384,
  "fcec1e5e06352a4b27e223f4dc64759483320da7": 1367,
  "fd294a381749153e97f7a56ea86b589ae39a5057": 25527,
  "fdb982a3222e71fd3a05d9c5f14d705c608681b8": 8923,
}
`;
